File: src/dere_ambient/triggers/priority.py
"""Priority calculation for curiosity items."""

from __future__ import annotations

from dataclasses import dataclass

from .types import CuriositySignal


@dataclass
class PriorityWeights:
    user_interest: float = 0.30
    knowledge_gap: float = 0.25
    type_weight: float = 0.20
    recency: float = 0.15
    exploration_count: float = 0.10


_TYPE_WEIGHTS = {
    "correction": 0.9,
    "emotional_peak": 0.7,
    "unfamiliar_entity": 0.5,
    "unfinished_thread": 0.6,
    "knowledge_gap": 0.6,
    "research_chain": 0.4,
}


def compute_curiosity_priority(
    signal: CuriositySignal,
    *,
    exploration_count: int = 0,
    recency: float = 1.0,
    weights: PriorityWeights | None = None,
) -> tuple[float, dict[str, float]]:
    weights = weights or PriorityWeights()
    type_weight = _TYPE_WEIGHTS.get(signal.curiosity_type, 0.5)

    exploration_boost = 1.0 if exploration_count <= 0 else max(0.0, 1.0 - 0.1 * exploration_count)

    factors = {
        "user_interest": _clamp(signal.user_interest),
        "knowledge_gap": _clamp(signal.knowledge_gap),
        "type_weight": _clamp(type_weight),
        "recency": _clamp(recency),
        "exploration_boost": _clamp(exploration_boost),
    }

    score = (
        weights.user_interest * factors["user_interest"]
        + weights.knowledge_gap * factors["knowledge_gap"]
        + weights.type_weight * factors["type_weight"]
        + weights.recency * factors["recency"]
        + weights.exploration_count * factors["exploration_boost"]
    )

    return _clamp(score), factors


def _clamp(value: float, minimum: float = 0.0, maximum: float = 1.0) -> float:
    return max(minimum, min(maximum, value))

File: packages/daemon/src/ambient-triggers/priority.ts
import type { CuriositySignal } from "./types.js";

export interface PriorityWeights {
  user_interest: number;
  knowledge_gap: number;
  type_weight: number;
  recency: number;
  exploration_count: number;
}

const DEFAULT_WEIGHTS: PriorityWeights = {
  user_interest: 0.3,
  knowledge_gap: 0.25,
  type_weight: 0.2,
  recency: 0.15,
  exploration_count: 0.1,
};

const TYPE_WEIGHTS: Record<string, number> = {
  correction: 0.9,
  emotional_peak: 0.7,
  unfamiliar_entity: 0.5,
  unfinished_thread: 0.6,
  knowledge_gap: 0.6,
  research_chain: 0.4,
};

export function computeCuriosityPriority(options: {
  signal: CuriositySignal;
  explorationCount?: number;
  recency?: number;
  weights?: PriorityWeights;
}): { score: number; factors: Record<string, number> } {
  const { signal, explorationCount = 0, recency = 1.0, weights = DEFAULT_WEIGHTS } = options;

  const typeWeight = TYPE_WEIGHTS[signal.curiosity_type] ?? 0.5;
  const explorationBoost =
    explorationCount <= 0 ? 1.0 : Math.max(0.0, 1.0 - 0.1 * explorationCount);

  const factors = {
    user_interest: clamp(signal.user_interest),
    knowledge_gap: clamp(signal.knowledge_gap),
    type_weight: clamp(typeWeight),
    recency: clamp(recency),
    exploration_boost: clamp(explorationBoost),
  };

  const score =
    weights.user_interest * factors.user_interest +
    weights.knowledge_gap * factors.knowledge_gap +
    weights.type_weight * factors.type_weight +
    weights.recency * factors.recency +
    weights.exploration_count * factors.exploration_boost;

  return { score: clamp(score), factors };
}

function clamp(value: number, minimum = 0, maximum = 1): number {
  return Math.max(minimum, Math.min(maximum, value));
}
